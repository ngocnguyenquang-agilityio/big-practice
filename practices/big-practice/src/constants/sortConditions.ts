export const SORT_CONDITIONS = [
  { label: 'High to Low', value: 'high-to-low' },
  { label: 'Low to High', value: 'low-to-high' },
];
