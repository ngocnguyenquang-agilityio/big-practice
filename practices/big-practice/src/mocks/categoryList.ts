export const mockCategoryList = [
  { value: 'phones', label: 'Phone' },
  { value: 'laptops', label: 'Laptop' },
];
