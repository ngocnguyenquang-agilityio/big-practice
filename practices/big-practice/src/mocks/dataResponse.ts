export const mockDataResponse = {
  products: [
    {
      id: 1,
      title: 'iPhone 9',
      price: 549,
      images: [
        'https://i.dummyjson.com/data/products/1/1.jpg',
        'https://i.dummyjson.com/data/products/1/2.jpg',
        'https://i.dummyjson.com/data/products/1/3.jpg',
        'https://i.dummyjson.com/data/products/1/4.jpg',
        'https://i.dummyjson.com/data/products/1/thumbnail.jpg',
      ],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/1/thumbnail.jpg',
      description: 'An apple mobile which is nothing like apple',
    },
    {
      id: 2,
      title: 'iPhone X',
      price: 899,
      images: [
        'https://i.dummyjson.com/data/products/2/1.jpg',
        'https://i.dummyjson.com/data/products/2/2.jpg',
        'https://i.dummyjson.com/data/products/2/3.jpg',
        'https://i.dummyjson.com/data/products/2/thumbnail.jpg',
      ],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/2/thumbnail.jpg',
      description:
        'SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip with ...',
    },
    {
      id: 3,
      title: 'Samsung Universe 9',
      price: 1249,
      images: ['https://i.dummyjson.com/data/products/3/1.jpg'],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/3/thumbnail.jpg',
      description: "Samsung's new variant which goes beyond Galaxy to the Universe",
    },
    {
      id: 4,
      title: 'OPPOF19',
      price: 280,
      images: [
        'https://i.dummyjson.com/data/products/4/1.jpg',
        'https://i.dummyjson.com/data/products/4/2.jpg',
        'https://i.dummyjson.com/data/products/4/3.jpg',
        'https://i.dummyjson.com/data/products/4/4.jpg',
        'https://i.dummyjson.com/data/products/4/thumbnail.jpg',
      ],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/4/thumbnail.jpg',
      description: 'OPPO F19 is officially announced on April 2021.',
    },
    {
      id: 5,
      title: 'Huawei P30',
      price: 499,
      images: [
        'https://i.dummyjson.com/data/products/5/1.jpg',
        'https://i.dummyjson.com/data/products/5/2.jpg',
        'https://i.dummyjson.com/data/products/5/3.jpg',
      ],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/5/thumbnail.jpg',
      description:
        'Huawei re-badged P30 Pro New Edition was officially unveiled yesterday in Germany and now the device has made its way to the UK.',
    },
    {
      id: 6,
      title: 'MacBook Pro',
      price: 1749,
      images: [
        'https://i.dummyjson.com/data/products/6/1.png',
        'https://i.dummyjson.com/data/products/6/2.jpg',
        'https://i.dummyjson.com/data/products/6/3.png',
        'https://i.dummyjson.com/data/products/6/4.jpg',
      ],
      category: 'laptops',
      thumbnail: 'https://i.dummyjson.com/data/products/6/thumbnail.png',
      description: 'MacBook Pro 2021 with mini-LED display may launch between September, November',
    },
    {
      id: 7,
      title: 'Samsung Galaxy Book',
      price: 1499,
      images: [
        'https://i.dummyjson.com/data/products/7/1.jpg',
        'https://i.dummyjson.com/data/products/7/2.jpg',
        'https://i.dummyjson.com/data/products/7/3.jpg',
        'https://i.dummyjson.com/data/products/7/thumbnail.jpg',
      ],
      category: 'laptops',
      thumbnail: 'https://i.dummyjson.com/data/products/7/thumbnail.jpg',
      description: 'Samsung Galaxy Book S (2020) Laptop With Intel Lakefield Chip, 8GB of RAM Launched',
    },
    {
      id: 8,
      title: 'Microsoft Surface Laptop 4',
      price: 1499,
      images: [
        'https://i.dummyjson.com/data/products/8/1.jpg',
        'https://i.dummyjson.com/data/products/8/2.jpg',
        'https://i.dummyjson.com/data/products/8/3.jpg',
        'https://i.dummyjson.com/data/products/8/4.jpg',
        'https://i.dummyjson.com/data/products/8/thumbnail.jpg',
      ],
      category: 'laptops',
      thumbnail: 'https://i.dummyjson.com/data/products/8/thumbnail.jpg',
      description:
        'Style and speed. Stand out on HD video calls backed by Studio Mics. Capture ideas on the vibrant touchscreen.',
    },
    {
      id: 9,
      title: 'Infinix INBOOK',
      price: 1099,
      images: [
        'https://i.dummyjson.com/data/products/9/1.jpg',
        'https://i.dummyjson.com/data/products/9/2.png',
        'https://i.dummyjson.com/data/products/9/3.png',
        'https://i.dummyjson.com/data/products/9/4.jpg',
        'https://i.dummyjson.com/data/products/9/thumbnail.jpg',
      ],
      category: 'laptops',
      thumbnail: 'https://i.dummyjson.com/data/products/9/thumbnail.jpg',
      description: 'Infinix Inbook X1 Ci3 10th 8GB 256GB 14 Win10 Grey â€“ 1 Year Warranty',
    },
  ],
  total: 10,
  skip: 0,
  limit: 9,
};

export const mockDataResponseTotal = {
  products: [
    {
      id: 1,
      title: 'iPhone 9',
      price: 549,
      images: [
        'https://i.dummyjson.com/data/products/1/1.jpg',
        'https://i.dummyjson.com/data/products/1/2.jpg',
        'https://i.dummyjson.com/data/products/1/3.jpg',
        'https://i.dummyjson.com/data/products/1/4.jpg',
        'https://i.dummyjson.com/data/products/1/thumbnail.jpg',
      ],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/1/thumbnail.jpg',
      description: 'An apple mobile which is nothing like apple',
    },
    {
      id: 2,
      title: 'iPhone X',
      price: 899,
      images: [
        'https://i.dummyjson.com/data/products/2/1.jpg',
        'https://i.dummyjson.com/data/products/2/2.jpg',
        'https://i.dummyjson.com/data/products/2/3.jpg',
        'https://i.dummyjson.com/data/products/2/thumbnail.jpg',
      ],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/2/thumbnail.jpg',
      description:
        'SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip with ...',
    },
    {
      id: 3,
      title: 'Samsung Universe 9',
      price: 1249,
      images: ['https://i.dummyjson.com/data/products/3/1.jpg'],
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/3/thumbnail.jpg',
      description: "Samsung's new variant which goes beyond Galaxy to the Universe",
    },
  ],
  total: 3,
  skip: 0,
  limit: 9,
};
