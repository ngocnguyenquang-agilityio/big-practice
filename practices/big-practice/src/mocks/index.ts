export * from './cart';
export * from './categoryList';
export * from './products';
export * from './sortConditions';
