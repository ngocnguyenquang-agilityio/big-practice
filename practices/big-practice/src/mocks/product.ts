export const mockProduct = {
  id: 1,
  title: 'Product 1',
  price: 20,
  category: 'phone',
  thumbnail:
    'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
};
