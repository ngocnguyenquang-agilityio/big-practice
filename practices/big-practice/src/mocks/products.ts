export const mockProduct = {
  id: 1,
  title: 'Product 1',
  price: 20,
  category: 'phone',
  thumbnail:
    'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
};

export const mockProducts = [
  {
    id: 1,
    title: 'Product 1',
    price: 10,
    category: 'phone',
    images: [
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
    ],
    thumbnail:
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
  },
  {
    id: 2,
    title: 'Product 2',
    price: 30,
    category: 'phone',
    images: [
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
    ],
    thumbnail:
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
  },
  {
    id: 3,
    title: 'Product 3',
    price: 20,
    category: 'phone',
    images: [
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
    ],
    thumbnail:
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
  },
];
