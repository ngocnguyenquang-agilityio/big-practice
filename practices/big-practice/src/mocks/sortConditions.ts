export const mockSortConditions = [
  { label: 'High to Low', value: 'high-to-low' },
  { label: 'Low to High', value: 'low-to-high' },
];
