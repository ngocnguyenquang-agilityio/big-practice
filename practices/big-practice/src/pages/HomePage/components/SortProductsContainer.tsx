// Libs
import { useCallback } from 'react';
import { useSearchParams } from 'react-router-dom';

// Components
import { Button } from '@components/Button';

type SortProducts = {
  sortCondition: Array<{ value: string; label: string }>;
  selectingItem: string;
  onSelectSort: (e: React.MouseEvent<HTMLButtonElement>) => void;
};

const SortProducts = ({ sortCondition = [], selectingItem, onSelectSort }: SortProducts): JSX.Element => {
  return (
    <nav>
      <h3 className='text-xs text-neutral-400 md:block'>Sort By</h3>
      <ul className='block'>
        {sortCondition.map((item) => (
          <li
            className={`mt-2 flex text-white ${item.value === selectingItem ? 'font-bold underline' : ''}`}
            key={item.value}
          >
            <Button
              variant='link'
              className='font-normal p-0 text-sm underline-offset-4 hover:underline hover:text-neutral-100'
              value={item.value}
              onClick={onSelectSort}
            >
              {item.label}
            </Button>
          </li>
        ))}
      </ul>
    </nav>
  );
};

const SortProductContainer = ({ sortConditions = [] }: { sortConditions: Array<{ label: string; value: string }> }) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const sortBy = searchParams.get('sort') || '';

  const handleSelectSort = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      searchParams.set('sort', e.currentTarget.value);
      setSearchParams(searchParams);
    },
    [searchParams],
  );

  return (
    <SortProducts
      selectingItem={sortBy}
      sortCondition={sortConditions}
      onSelectSort={handleSelectSort}
    />
  );
};

export default SortProductContainer;
