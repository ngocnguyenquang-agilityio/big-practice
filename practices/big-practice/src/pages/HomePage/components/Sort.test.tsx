// Libs
import { MemoryRouter } from 'react-router-dom';
import { screen, render, RenderResult, fireEvent } from '@testing-library/react';

// Components
import { SortProducts } from './SortProductsContainer';

// Mocks
import { mockSortConditions } from '../../../mocks/sortConditions';

const props = {
  sortCondition: mockSortConditions,
  selectingItem: 'low-to-high',
  onSelectSort: jest.fn(),
};

describe('Sort Product component', () => {
  let comp: RenderResult;
  beforeEach(() => {
    comp = render(
      <MemoryRouter>
        <SortProducts {...props} />
      </MemoryRouter>,
    );
  });

  test('Should render correctly', () => {
    expect(comp).toBeTruthy();
  });

  test('Should match snapshot', () => {
    expect(comp).toMatchSnapshot();
  });

  test('Should render correct number of items', async () => {
    const items = await screen.getByTestId('sort-products');
    expect(items.children).toHaveLength(2);
  });

  test('Should has no underline item in first render', async () => {
    const items = await screen.getByTestId('sort-products');
    expect(items.getElementsByClassName('font-bold underline').length).toBe(1);
  });

  test('Should call onSelectSort', async () => {
    await screen.getByTestId('sort-products');
    const button = screen.getByRole('button', { name: /Low to High/i });
    fireEvent.click(button);
    expect(props.onSelectSort).toBeCalledTimes(1);
  });
});
