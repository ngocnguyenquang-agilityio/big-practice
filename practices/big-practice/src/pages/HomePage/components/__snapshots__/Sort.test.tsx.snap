// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Product component Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav>
        <h3
          class="hidden text-xs text-neutral-400 md:block"
        >
          Sort By
        </h3>
        <ul
          class="hidden md:block"
          data-testid="sort-products"
        >
          <li
            class="mt-2 flex text-white "
          >
            <button
              class="text-sm underline-offset-4 hover:underline hover:text-neutral-100"
              value="high-to-low"
            >
              High to Low
            </button>
          </li>
          <li
            class="mt-2 flex text-white font-bold underline"
          >
            <button
              class="text-sm underline-offset-4 hover:underline hover:text-neutral-100"
              value="low-to-high"
            >
              Low to High
            </button>
          </li>
        </ul>
        <div
          class="md:hidden"
        >
          <div
            class="relative"
          >
            <select
              class="bg-transparent flex w-full items-center justify-between rounded border px-4 py-2 text-sm border-white/30"
            >
              <option
                class="mt-2 flex text-white bg-black"
                value="high-to-low"
              >
                High to Low
              </option>
              <option
                class="mt-2 flex text-white bg-black"
                value="low-to-high"
              >
                Low to High
              </option>
            </select>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav>
      <h3
        class="hidden text-xs text-neutral-400 md:block"
      >
        Sort By
      </h3>
      <ul
        class="hidden md:block"
        data-testid="sort-products"
      >
        <li
          class="mt-2 flex text-white "
        >
          <button
            class="text-sm underline-offset-4 hover:underline hover:text-neutral-100"
            value="high-to-low"
          >
            High to Low
          </button>
        </li>
        <li
          class="mt-2 flex text-white font-bold underline"
        >
          <button
            class="text-sm underline-offset-4 hover:underline hover:text-neutral-100"
            value="low-to-high"
          >
            Low to High
          </button>
        </li>
      </ul>
      <div
        class="md:hidden"
      >
        <div
          class="relative"
        >
          <select
            class="bg-transparent flex w-full items-center justify-between rounded border px-4 py-2 text-sm border-white/30"
          >
            <option
              class="mt-2 flex text-white bg-black"
              value="high-to-low"
            >
              High to Low
            </option>
            <option
              class="mt-2 flex text-white bg-black"
              value="low-to-high"
            >
              Low to High
            </option>
          </select>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
